Update diff gutter on document save

jeeeesper opened the pull request:
This relates to #4994 but does not try to add any file watching or such. This simple commit triggers an update of the diff base when saving a document, and also allowing that base to be unset, if the new document path does not exist yet.

For me, this came up when I was saving a file under a different path, but the diff gutter would still show the changes with respect to the original file at the HEAD revision.

astralaster commented:
This does not work correctly yet. If you create a new file and add it to git, I would expect the gutter to update on edit / write then, but it does only update if you reload the file.

jeeeesper commented:
That is correct, I did not realise that the function returns early if the path does not change. I updated the commit.

I have now also added that the document language is re-detected on save even if the path does not change. This (to my knowledge) is only then relevant when the shebang changes. We can move that to another issue / PR though, if needed.

warmwaffles commented:
I have a key mapping that I mash after I save to reload the document. It's a crappy work around, but it's how I get that cleaned up today.
